hazelcast/src/main/java/com/hazelcast/map/impl/query/Target.java

```diff
@@ -79,17 +81,18 @@ public int getClassId() {
 
     @Override
     public void writeData(ObjectDataOutput out) throws IOException {
-        out.writeInt(partitionId);
+        // TODO use IDS
```

Comment:
Changing the serialized form of a class in a patch release will break patch-level-compatibility: 4.1.1 will not be able to participate in the same cluster as 4.1 member. The class evolution support we have for rolling upgrades doesn't help much here, because it is geared towards minor-version compatibility.

In terms of serialization, it would be probably ok to start with the same format (1 `int`, 1 `UTF` field), then append the `partitions` object. This way, a 4.1 receiver will just ignore whatever is past the 2 fields it is already aware of, and not throw a serialization exception. The other way around, a 4.1.1 receiver of a `Target` from a 4.1 member can try to read the `partitions` object after the `UTF` field and expect an exception (assuming then that it was sent from a 4.1 member). However this only solves the "do not throw a serialization exception" part of the problem. The semantics of deserialized object would be the real challenge to solve. Is there any way to avoid changing `Target`'s serialized form?

Reply:
Hmm, didn't realize this. But I think the class is never serialized. It's only used in the `MapProxySupport` and `QueryExecutionEngine` locally to send the right operations to right members. I'm [trying this](https://github.com/hazelcast/hazelcast/pull/17886/commits/11c5050b0671632f97acda8376871ff71c21005b), see if any tests fail.

Reply:
I made the `Target` non-serializable. IMO it's pretty easy to reason from the code that it's never serialized: It's just a value that always lands in `QueryEngine.execute` where it is used to create operations, but it's not a part of them. I also checked the enterprise code, there's no another implementation of `QueryEngine`.